core/src/smartcontracts/wasm/codec/derive/src/lib.rs

```diff
@@ -0,0 +1,387 @@
+//! This crate provides [`wrap`] attribute macro to wrap a host-defined function into another
+//! function which signature will be compatible with `wasmtime` crate to be successfully exported.
+
+use std::ops::Deref;
+
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
+use proc_macro_error::{abort, diagnostic, proc_macro_error, Diagnostic, Level, OptionExt as _};
+use quote::quote;
+use syn::{parse_quote, punctuated::Punctuated};
+
+/// Macro to wrap function with normal parameters and return value to another one which will
+/// feet `wasmtime` specifications.
+///
+/// Describing all possible input and output signatures would be a very big table,
+/// so see detailed signature by expanding generated code (i.e. with `cargo expand`).
+///
+/// # Key notes
+///
+/// 1. If there is something to (de-)code (input or output) generated signature will always
+/// return `Result<..., Trap>`.
+/// 2. If your function returns `T` on success, then generated function will
+/// return `Result<WasmUsize, Trap>`, where `WasmUsize` is the offset of encoded `T`
+/// 3. If your function returns [`Result`] with `Trap` on [`Err`], generated function will pop it up
```

Comment:
I also thought so at the beginning. But then I realized that users may also want to return *unrecoverable errors from smart contract point of view*. 
his is done for `log` function where user can submit inexistent level number.

Reply:
but user should never return `Trap`. Codec should modify the function so that it returns trap?

Reply:
check my other comment. `log` function should panic instead IMO

Reply:
I thought that if we panic -- then it will be a unrecoverable error on Iroha side. And if we return `Trap` -- it will be an unrecoverable error on wasm side. No?

Reply:
yes, but not quite. If you `panic!` inside the rust program it's as if you have returned `Trap` explicitly. Therefore, any function that returns `Result<_, Trap>` can be transformed into an equivalent function that panics. 

Our functions should never return `Trap` explicitly, this should be handled by your `codec::wrap`

Reply:
We decided to implement it in another task: #3555